src/selector-observer.js

```diff
@@ -0,0 +1,194 @@
+import { adoptCSS, unadoptCSS } from "./util/adopt-css.js";
+
+/**
+ * @typedef {Object} SelectorObserverOptions
+ * @property {(Document | ShadowRoot)[]} [roots=[document]] - The roots to observe.
+ * @property {boolean} [any] - Whether to fire the callback when any of the observed selectors are matched or when individual selectors are matched.
+ */
+
+export default class SelectorObserver {
+	/**
+	 * Map of root nodes to sheets
+	 * @type {Map<Document | ShadowRoot, CSSStyleSheet>}
+	 */
+	#roots = new Map();
+
+	#selectors = new Set();
+
+	/**
+	 * @param {Function} callback
+	 * @param {SelectorObserverOptions} [options]
+	 */
+	constructor (callback, options = {}) {
+		this.callback = callback;
+		this.options = options;
+
+		if (this.options.roots) {
+			for (let root of this.options.roots) {
+				this.observeRoot(root);
+			}
+		}
+		else {
+			this.observeRoot();
+		}
+	}
+
+	/**
+	 * Observe all selectors on a new root node.
+	 */
+	observeRoot (root = document) {
+		if (this.#roots.has(root)) {
+			return;
+		}
+
+		this.#roots.set(root, null);
+		this.#updateCSS(root);
+
+		root.addEventListener("transitionend", this, { capture: true });
+		root.addEventListener("animationend", this, { capture: true });
+	}
+
+	handleEvent (event) {
+		let { target } = event;
+		if (event.type === "transitionend") {
+			let selector = decode(event.propertyName.replace(/^--selector-observer-selector-/, ""));
+
+			console.log("transitionend", event.target, selector);
+			// this.callback(event.target);
+		}
+		else if (event.type === "animationend") {
+			let selector = decode(event.animationName.replace(/^selector-observer-/, ""));
+			console.log("animationend", event.target, selector);
+			// this.callback(event.target);
+		}
+	}
+	/**
+	 * Stop observing all selectors on a root node.
+	 */
+	unobserveRoot (root = document) {
+		if (!this.#roots.has(root)) {
+			return;
+		}
+
+		this.#roots.delete(root);
+
+		let sheet = this.#roots.get(root);
```

Comment:
I thought we had already removed the `root` from `roots` (in the line above) and can't get the style sheet associated with it, no? Shouldn't we move `this.#roots.delete(root)` at the end of `unobserveRoot`?

Reply:
Gah, you are 100% right